### DISCUSSION TITLE
Need help as I am trying to loop backwards through the IN1 segments in my outbound messages and delete any segment where IN1.2.1 is not an acceptable code

### DISCUSSION THREAD

**Discussion by mayowan16** [nextgenhealthcare/connect#6346] [Q&A] [answered]
We have 36 insurance plans that we don't want crossing our interface.  I am trying to loop through the IN1 segments backwards, then delete any corresponding segment that contains one of the 36 unacceptable values.  I end up just deleting all of the IN1 segments or none of them delete.  I am not sure how to combine the reverse looping of the basic IN1 segments with the IF variable found in IN1.2.1.  Any help would be appreciated.

```
var i = 0

var in12 = msg['IN1'][i]['IN1.2']['IN1.2.1'].toString(); 

for (var i = msg['IN1'].index + 5 ; i > -1; i--) {								
	if (in12 == 'C0037' || in12 == 'C0095' || in12 == 'C0104' || in12 == 'C0116' || in12 == 'C0146' ||
	    in12 == 'C0155' || in12 == 'C0166' || in12 == 'C0212' || in12 == 'C0253' || in12 == 'C0269' ||
	    in12 == 'C0294' || in12 == 'C0332' || in12 == 'C0356' || in12 == 'C0357' || in12 == 'C0389' ||
	    in12 == 'C0409' || in12 == 'C0487' || in12 == 'C0503' || in12 == 'C0523' || in12 == 'C0524' ||
	    in12 == 'C0529' || in12 == 'C0538' || in12 == 'C0548' || in12 == 'C0549' || in12 == 'MC020' ||
	    in12 == 'MC021' || in12 == 'MC022' || in12 == 'MC023' || in12 == 'MC027' || in12 == 'MC034' ||
	    in12 == 'MC115' || in12 == 'MC116' || in12 == 'MC181' || in12 == 'MC182' || in12 == 'MC202' || in12 == 'MC204') {
		delete msg['IN1'][i];	
	}
}
```

**Comment by pacmano1** [comment]
The first obvious thing is your assignment used for comparison is outside the for loop.

**Comment by mayowan16** [comment]
Hi there, thank you for that response.  I have tried nesting the assignments inside the loop.  Now, I am getting the following error.  TypeError: Cannot read property "IN1.2" from undefined.  I'm not sure what else I need to do here.
``for (var i = 0; i < 6; i--) {	
	const in12 = msg['IN1'][i]['IN1.2']['IN1.2.1'].toString(); }
	{							
	if (in12 == 'C0037' || in12 == 'C0095' || in12 == 'C0104' || in12 == 'C0116' || in12 == 'C0146' ||
	    in12 == 'C0155' || in12 == 'C0166' || in12 == 'C0212' || in12 == 'C0253' || in12 == 'C0269' ||
	    in12 == 'C0294' || in12 == 'C0332' || in12 == 'C0356' || in12 == 'C0357' || in12 == 'C0389' ||
	    in12 == 'C0409' || in12 == 'C0487' || in12 == 'C0503' || in12 == 'C0523' || in12 == 'C0524' ||
	    in12 == 'C0529' || in12 == 'C0538' || in12 == 'C0548' || in12 == 'C0549' || in12 == 'MC020' ||
	    in12 == 'MC021' || in12 == 'MC022' || in12 == 'MC023' || in12 == 'MC027' || in12 == 'MC034' ||
	    in12 == 'MC115' || in12 == 'MC116' || in12 == 'MC181' || in12 == 'MC182' || in12 == 'MC202' || in12 == 'MC204') {
		delete msg['IN1'][i];	
	}
}``

**Comment by pacmano1** [comment, marked as answer]
One of several ways but in this case we are also checking to make sure there is an IN1.2.1.

I will leave the re-order of the set id to you.  Assumes a more recent version of mirth for the .includes.

```JS
/**
 * $t is an inline try/catch to take the place of optional chaining ie var d = $t(()=> a.b.c) is the same as var d = a?.b?.c
 * Check if $t is already defined. If not then define it.
 * @Author Michael L Hobbs
 * @History
 * 20220326 init
 */
if (typeof $t === 'undefined') {
    function $t(cb) {
        try {
            return cb()
        } catch (ignore) {
            // do nothing
        }
    }
}

var valuesToFilter = [
    "C0037", "C0095", "C0104", "C0116", "C0146",
    "C0155", "C0166", "C0212", "C0253", "C0269",
    "C0294", "C0332", "C0356", "C0357", "C0389",
    "C0409", "C0487", "C0503", "C0523", "C0524",
    "C0529", "C0538", "C0548", "C0549", "MC020",
    "MC021", "MC022", "MC023", "MC027", "MC034",
    "MC115", "MC116", "MC181", "MC182", "MC202", "MC204"
];


for (var i = msg['IN1'].length(); i >= 0; i--) {
	var this_in1 = $t(() => msg['IN1'][i]['IN1.2']['IN1.2.1'].toString()) || null
	if (this_in1 != null && valuesToFilter.includes(this_in1)) delete msg['IN1'][i]
}
```

this for each would work also instead of the above for loop but with the rest of the code of course.

```JS
for each (in1 in msg.IN1) {
	var this_in1 = $t(() => in1['IN1.2']['IN1.2.1'].toString()) || null
	if (this_in1 != null && valuesToFilter.includes(this_in1)) delete in1[0]
}
```
